### DISCUSSION TITLE
Look ahead bias of CryptoTradingEnv in Chapter 4

### DISCUSSION THREAD

**Discussion by QuantHao** [PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook#39] [Reinforcement Learning] [answered]
Hi, @praveen-palanisamy 

In Quant finance, look ahead bias in training and back-test is a often pitfall. When you create CryptoTradingEnv, `self.ohlcv_df` is sorted by date _**ascending**_ https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/31f8376095e635ab492bd853628417b9ef82694e/Chapter04/crypto_trading_env.py#L58-L62

Observation token each step is
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/31f8376095e635ab492bd853628417b9ef82694e/Chapter04/crypto_trading_env.py#L138-L141

I think `self.current_step : self.current_step + self.horizon` might cause look-ahead bias. Do you think change into `self.current_step - self.horizon : self.current_step` might be better?

**Comment by QuantHao** [comment]
Same issue applies to Chapter 5.

**Comment by praveen-palanisamy** [comment, marked as answer]
>In Quant finance, look ahead bias in training and back-test is a often pitfall.

Good point!

The `current_step` used in this RL context is just an abstraction for the MDP step number. There's really no representation for _current_ (time) step. So, the `horizon` used in the above recipe in practical terms is equivalent to the `history` and not the _future_ / `look-ahead`.

For example, 
Assuming a value of `30` for the horizon as per the default config in the recipe:
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/31f8376095e635ab492bd853628417b9ef82694e/Chapter04/crypto_trading_env.py#L21-L22

The agent would observe the ticker data for the _past_ 30 days/hours/minutes in one observation. Practically, when using a live-stream of ticker data (from an exchange server), we would buffer 30 data points before we can start getting action predictions from the agent.

>I think self.current_step : self.current_step + self.horizon might cause look-ahead bias. 

No. This doesn't mean the agent is really looking-ahead. The intended/correct interpretation is to treat the horizon as _historical_ data fed to the agent (not future/look-ahead data).

>Do you think change into self.current_step - self.horizon : self.current_step might be better?

Both are equivalent. I made the choice to use the first option since that's simpler to represent in a data frame and matches real data streams (ascending order in time) from  exchanges.

Hope that addresses your concern?

**Comment by QuantHao** [comment]
I see. Thanks for answering.
